# 단계 설정

> [출처](https://www.daleseo.com/github-actions-steps/)

## 불안정한 단계의 실패 무시

작업(job) 실행 도중 단계(step)이 실패하면 그 이후의 단계는 실행되지 않고 중단된다. 합리적으로 여겨짐.

하지만 실제 플젝에서 성패가 오락가락하는 불안정한 단계가 있을 수 있다. 대표적인 예로 테스트 케이스가 실패하는 경우다. 테스트 단계가 실패할 때마다 작업 전체가 중단된다면 팀 전체가 곤란해짐.

단계는 `continue-on-error` 속성을 지원함. 이 속성을 `true`로 지정할 경우 단계가 실패 하더라도 작업은 중단되지 않고 남은 단계를 계속해서 실행해준다.

예를 덜어, 다음 작업에서 첫 번째 단계는 항상 실패하게 되지만 `continue-on-error` 속성이 `true` 로 되어 있기 때문에 두 번째 단계에서 `I don't care!`가 출력된다.

```yaml
name: Our Steps
on: push
jobs:
  ignore:
    runs-on: ubuntu-latest
    steps:
      - id: flaky
        continue-on-error: true
        run: exit 1
      - run: echo "I don't care!"
```

<br/>

## 이전 단계의 성패 여부와 상관없이 다음 단계 수행

만약 이전 단계 성패 여부와 상관없이 무조건 수행되어야 하는 단계가 있다면 어떻게 해야할까? 흔한 사례로 작업의 실행 결과를 이메일이나 메세징 어플리케이션으로 통보해야할 때.

이럴 경우 무조건 수행되어야 하는 단계의 `if` 속성에 `always()` 라는 표현식을 설정해주면 된다.

예를 들어, 첫 번째 단계를 랜덤하게 성공하거나 실패하게 한 다음, 두 번째 단계에서 항상 첫 번째 단계의 결과가 출력되도록 작업 설정을 해보자. 특정 단계의 출력 결과를 확인하기 위해서 `steps.<step_id>.outcome` 이라는 Github Actions의 문맥(context)을 사용하고 있다.

```yaml
name: Our Steps
on: push
jobs:
  notify:
    runs-on: ubuntu-latest
    steps:
      - id: random
        run: if [[ $(($RANDOM % 2)) == 0 ]]; then exit 0; else exit 1; fi
      - if: ${{ always() }}
        run: echo ${{ steps.random.outcome }}
```

만약 첫 번째 단계가 성공했다면, 두 번재 단계에서 `success`가 콘솔에 출력될 것이다. 하지만 첫 번째 단계가 실패했다면, 두 번째 단계에서 `failure`가 콘솔에 출력될 것임.

<br/>

## 이전 단계가 실패했을 때만 다음 단계 수행

어떤 단계가 실패했을 때만 예비로 수행될 백업(backup) 단계를 설정해야될 때가 있음. `if` 속성의 `failure()` 표현식을 설정하면 된다.

```yaml
name: Our Steps
on: push
jobs:
  backup:
    runs-on: ubuntu-latest
    steps:
      - name: original
        run: exit 1
      - name: backup
        if: ${{ failure() }}
        run: echo backup
```

